addons/sale_timesheet/models/project.py

```diff
@@ -300,37 +301,46 @@ def _get_profitability_items(self):
-            'margin': profitability and profitability[0]['margin'] or 0.0,
-            'revenues': profitability and (profitability[0]['amount_untaxed_invoiced'] + profitability[0]['amount_untaxed_to_invoice'] +
-                                           profitability[0]['expense_amount_untaxed_invoiced'] + profitability[0]['expense_amount_untaxed_to_invoice'] +
-                                           profitability[0]['other_revenues']) or 0.0,
-            'timesheet_unit_amount': profitability and self._convert_project_uom_to_timesheet_encode_uom(profitability[0]['timesheet_unit_amount']) or 0.0,
-            'timesheet_percentage_billable': timesheet_billable and timesheet_billable / (timesheet_billable + timesheet_non_billable) * 100 or 0.0,
-        }
+        result = {}
+        if costs_revenues:
+            profitability = self.env['project.profitability.report'].read_group(
+                [('project_id', '=', self.id)],
+                ['project_id',
+                 'amount_untaxed_to_invoice',
+                 'amount_untaxed_invoiced',
+                 'expense_amount_untaxed_to_invoice',
+                 'expense_amount_untaxed_invoiced',
+                 'other_revenues',
+                 'expense_cost',
+                 'timesheet_cost',
+                 'margin'],
+                ['project_id'])
+            result.update({
+                'costs': profitability and profitability[0]['timesheet_cost'] + profitability[0]['expense_cost'] or 0.0,
+                'margin': profitability and profitability[0]['margin'] or 0.0,
+                'revenues': profitability and (profitability[0]['amount_untaxed_invoiced'] + profitability[0]['amount_untaxed_to_invoice'] +
+                                               profitability[0]['expense_amount_untaxed_invoiced'] + profitability[0]['expense_amount_untaxed_to_invoice'] +
+                                               profitability[0]['other_revenues']) or 0.0,
+            })
+        else:
+            result.update(costs=0.0, margin=0.0, revenues=0.0)
+        if timesheets:
+            timesheets = self.env['account.analytic.line'].read_group([('project_id', '=', self.id)], ['so_line', 'unit_amount'], ['so_line'])
+            timesheet_billable = timesheet_non_billable = 0.0
+            for timesheet in timesheets:
+                if timesheet['so_line']:
+                    timesheet_billable += timesheet['unit_amount']
+                else:
+                    timesheet_non_billable += timesheet['unit_amount']
+            result.update({
+                'timesheet_unit_amount': self._convert_project_uom_to_timesheet_encode_uom(timesheet_billable + timesheet_non_billable),
+                'timesheet_percentage_billable': timesheet_billable and timesheet_billable / (timesheet_billable + timesheet_non_billable) * 100 or 0.0,
+            })
+        else:
+            result.update(timesheet_unit_amount=0.0, timesheet_percentage_billable=0.0)
```

Comment:
Since when `timesheets=False` and `costs_revenues` all values in the result dict will be equal to 0.0, we could create the result with all keys and their value is equal to 0.0. So, we avoid 2 useless `update`. 🙂

For the first read_group about `cost_revenues`, we could add `limit=1` since we take only the first element in the `read_group` result. What do you think ?

Feel free to update my suggestion, I think the profitability.get(..., 0.0) is not necessary and we can do as you have done.

```suggestion
        result = {
            'costs': 0.0,
            'margin': 0.0,
            'revenues': 0.0,
            'timesheet_unit_amount': 0.0,
            'timesheet_percentage_billable': 0.0,
        }
        if costs_revenues:
            profitability = self.env['project.profitability.report'].read_group(
                [('project_id', '=', self.id)],
                ['project_id',
                 'amount_untaxed_to_invoice',
                 'amount_untaxed_invoiced',
                 'expense_amount_untaxed_to_invoice',
                 'expense_amount_untaxed_invoiced',
                 'other_revenues',
                 'expense_cost',
                 'timesheet_cost',
                 'margin'],
                ['project_id'])  # XBO: add limit=1 in the read_group ?
            if profitability:
                profitability = profitability[0]
                result.update({
                    'costs': profitability.get('timesheet_cost', 0.0) + profitability.get('expense_cost', 0.0),
                    'margin': profitability.get('margin', 0.0),
                    'revenues':
                        profitability.get('amount_untaxed_invoiced', 0.0)
                        + profitability.get('amount_untaxed_to_invoice', 0.0)
                        + profitability.get('expense_amount_untaxed_invoiced', 0.0)
                        + profitability.get('expense_amount_untaxed_to_invoice', 0.0)
                        + profitability.get('other_revenues', 0.0),
                })
        if timesheets:
            timesheets = self.env['account.analytic.line'].read_group([('project_id', '=', self.id)], ['so_line', 'unit_amount'], ['so_line'])
            timesheet_billable = timesheet_non_billable = 0.0
            for timesheet in timesheets:
                if timesheet['so_line']:
                    timesheet_billable += timesheet['unit_amount']
                else:
                    timesheet_non_billable += timesheet['unit_amount']
            result.update({
                'timesheet_unit_amount': self._convert_project_uom_to_timesheet_encode_uom(timesheet_billable + timesheet_non_billable),
                'timesheet_percentage_billable': timesheet_billable and timesheet_billable / (timesheet_billable + timesheet_non_billable) * 100 or 0.0,
            })
```

Reply:
Done with some adaptation of your suggestion, as you suggested it in your last sentence :D